### PR TITLE
#841 adding vvn and fridge columns

### LINKED ISSUE #841: [Vaccine Module][Supplier Invoice] - Add VVM and Fridge assignment column
As per epic: https://github.com/sussol/org-issues/issues/23

#### Description

Two columns to be added to Supplier Invoice, they should only be rendered if at least one item in invoice 'isVaccine'

If at this stage item does not already have a getter isVaccine, create one, it should safely check for item.categroy.name = 'Vaccine'.

If at this stage location does not have isFridge getter, create one, it should safely check for location.locationType.description = 'fridge' (ignore case, so turn to lower/upper case)

##### Fridge Selection

Default (first for now) fridge should be assigned to transactionBatches for all vaccines items. (Should this be done when page loads and when new vaccine item is added).

Modal pops up with all fridges when fridge cell is touched

##### VVM Status

use this component: https://github.com/openmsupply/mobile/issues/840 

When this component is toggled to false a modal should pop up asking what quantity to apply this to. The transactionBatch should be split accordingly.

###### DB Logic

Make sure itemBatch inherits transactionBatch.location 

Inventory adjustment to remove VVM fails from stock and record VVM passes should be created in seperate issue (https://github.com/openmsupply/mobile/issues/842)

#### Design

Had comments on the design, saying 'underscore look like a link, should have chevron'
fridge name + chevron instead of just underscore.

VM - Supplier Invoice No Toggle (1)-1

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #841 

image

## Requirements from issue
- Two columns to be added to Supplier Invoice, they should only be rendered if at least one item in invoice 'isVaccine'
    - `componentDidMount` uses a new helper to check the transaction for a vaccine. If a vaccine is added with `onNewItem` set the `hasVaccine` state to true [used state here for performance rather than calling the transaction method every time to check - not sure if best]
    - Both columns have been added, not sure if the sizing is correct.
    - Using a slightly more dynamic method to get the columns to use - not sure if I like it
-  If at this stage item does not already have a getter isVaccine, create one, it should safely check for item.categroy.name = 'Vaccine'.
-  If at this stage location does not have isFridge getter, create one, it should safely check for location.locationType.description = 'fridge' (ignore case, so turn to lower/upper case)
-  Default (first for now) fridge should be assigned to transactionBatches for all vaccines items.
    - Done on creating a batch and on mounting. Maybe a `get vaccines` helper method for `Transaction` could be useful?
-  Modal pops up with all fridges when fridge cell is touched
-  Make sure itemBatch inherits transactionBatch.location
    - I think this is on finalizing, which will be done in a different PR. Although I have the question, what happens when an `itemBatch` has multiple `transactionBatch` with different `locations`
-  Had comments on the design, saying 'underscore look like a link, should have chevron'
fridge name + chevron instead of just underscore.
    - Used a chevron and a 'X' with words for failing VVM status locations - not sure if I like this, but I kept trying to click the fridge

At the moment:
- Finalizing not in, will do in another (inventory adjustments etc)
- All locations are used in the modal. I will use `GenericChooseModal` once IC stock takes PR is merged into develop. (Can rebase `vaccine-development`?), but at the moment `AutoCompleteSelector` only takes a `Realm.results` also.
- I want to make a new component for the dosage entering modal. `GenericNumberEntering` or something which has validation and feedback for the user on the rare use case they try to enter non-numeric characters. Maybe some + and - buttons?

### CODE REVIEW

File: src/pages/SupplierInvoicePage.js [hunk 1]
```diff
@@ -137,12 +248,20 @@ export class SupplierInvoicePage extends React.Component {
...
-      createRecord(
+      const transactionBatch = createRecord(
         database,
         'TransactionBatch',
         transactionItem,
         createRecord(database, 'ItemBatch', item, '')
       );
+
+      if (item.isVaccine) {
+        transactionBatch.location = database
```
Comment (andreievg) [inline src/pages/SupplierInvoicePage.js]: i think it's safe to store default fridge in this.defaultFridge ?

File: src/pages/SupplierInvoicePage.js [hunk 2]
```diff
@@ -225,15 +352,42 @@ export class SupplierInvoicePage extends React.Component {
...
+            leftText="PASS"
+            rightText="FAIL"
+            currentState={isVVMPassed}
+            onPress={this.onVvmToggle(transactionBatch)}
+          />
+        );
+      }
+      case 'fridge': {
+        if (!isVaccine) return emptycell;
+        const failed = isVVMPassed === false;
```
Comment (andreievg) [inline src/pages/SupplierInvoicePage.js]: should we show 'no fridges' <- if no default fridge ?

File: src/database/DataTypes/TransactionBatch.js [hunk 3]
```diff
@@ -29,8 +29,15 @@ export class TransactionBatch extends Realm.Object {
...
 
     // Can safely remove associated item batch if transaction batch was created by an external
-    // supplier invoice.
-    if (this.transaction.isExternalSupplierInvoice) {
+    // supplier invoice unless it is a vaccine, which are split in a supplier invoice if the
+    // VVM status has failed. If the itemBatch has more than one TransactionBatch, there is
+    // another transactionBatch associated so don't delete.
+    let canDeleteItemBatch = true;
+    if (this.isVaccine) {
+      canDeleteItemBatch = this.itemBatch.transactionBatch.length === 1;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionBatch.js]: done 5c422acb8ca4724615e673fefcaf578a81589760

File: src/database/DataTypes/TransactionBatch.js [hunk 4]
```diff
@@ -145,6 +160,14 @@ export class TransactionBatch extends Realm.Object {
     return quantity;
   }
 
+  get transactionItem() {
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionBatch.js]: Forgot to comment in the PR. I thought I needed this, I didn't. Written now so might leave it unless I should just remove it

File: src/database/DataTypes/TransactionBatch.js [hunk 5]
```diff
@@ -153,6 +176,34 @@ export class TransactionBatch extends Realm.Object {
...
+   * from within a supplier invoice - expiryDate and location.
+   * @param {Realm}  database
+   * @param {Number} splitValue - the value of numberOfPacks to split the batches on
+   */
+  splitBatch({ database, splitValue }) {
+    const { itemBatch, transactionItem, numberOfPacks } = this;
+
+    let newTransactionBatch;
+    database.write(() => {
+      newTransactionBatch = createRecord(database, 'TransactionBatch', transactionItem, itemBatch);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionBatch.js]: Merged in. Much nicer way of doing this. I think it seems a little awkward either way [new batch being failed or old]. Thanks!

Added this commit to it afa81d9a2fd2c7be3158edbad2dc89d211d4ac10

File: src/pages/SupplierInvoicePage.js [hunk 6]
```diff
@@ -282,12 +436,37 @@ export class SupplierInvoicePage extends React.Component {
             }}
           />
         );
+      case FRIDGE_SELECT:
+        return (
+          <AutocompleteSelector
+            options={database.objects('Location')}
```
Comment (andreievg) [inline src/pages/SupplierInvoicePage.js]: i guess you are planning to change this (to make sure location is fridge) when the new modal with array input is available ?

File: src/pages/SupplierInvoicePage.js [hunk 7]
```diff
@@ -298,10 +477,25 @@ export class SupplierInvoicePage extends React.Component {
...
+    const vaccineColumns = [
+      'itemCode',
+      'itemName',
+      'totalQuantity',
+      'expiryDate',
+      'vvm',
+      'fridge',
+      'remove',
+    ];
+    const normalColumns = ['itemCode', 'itemName', 'totalQuantity', 'expiryDate', 'remove'];
```
Comment (andreievg) [inline src/pages/SupplierInvoicePage.js]: nice way to do columns!

File: src/database/DataTypes/TransactionBatch.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/database/DataTypes/TransactionBatch.js]: @joshxg, don't really want to have something that's not used. if you want to keep it:
return transaction.items.filtered('batches.id = $0', this.id)

File: src/pages/SupplierInvoicePage.js (hunk 1 above)
Comment (joshxg) [reply src/pages/SupplierInvoicePage.js]: Done - 4b4fdc92eb3230c3b32898fa188526581e579061

File: src/pages/SupplierInvoicePage.js (hunk 2 above)
Comment (joshxg) [reply src/pages/SupplierInvoicePage.js]: done 0908a8201be02fa258d976dd99b9201dd7e2f1a2

3 states for fridge cell: 'No fridges'/'Discarded' (both using the same style, with a cross and low opacity gray) and pressable 'Fridge X (name)' with a caret

File: src/pages/SupplierInvoicePage.js (hunk 6 above)
Comment (joshxg) [reply src/pages/SupplierInvoicePage.js]: Yeah. From my PR comment. I should make them shorter :(

At the moment:

    Finalizing not in, will do in another (inventory adjustments etc)
    All locations are used in the modal. I will use GenericChooseModal once IC stock takes PR is merged into develop. (Can rebase vaccine-development?), but at the moment AutoCompleteSelector only takes a Realm.results also.
    I want to make a new component for the dosage entering modal. GenericNumberEntering or something which has validation and feedback for the user on the rare use case they try to enter non-numeric characters. Maybe some + and - buttons?

**Pull request merged**